xml_struct/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+//!
+//! In recognition of these limitations, this crate should not be published to
+//! crates.io at this time. If a generalized implementation generates interest
+//! or is thought to have merit, these limitations may be addressed at a later
+//! time.
+
+mod impls;
+mod tests;
+
+use quick_xml::{
+    events::{BytesEnd, BytesStart, Event},
+    Writer,
+};
+use thiserror::Error;
+
+pub use xml_struct_derive::*;
+
+/// A data structure which can be serialized as XML content nodes.
+///
+/// # Usage
+///
+/// The following demonstrates end-to-end usage of `XmlSerialize` with both
+/// derived and manual implementations.
+///
+/// ```
+/// use quick_xml::{
+///     events::{BytesText, Event},
+///     writer::Writer
+/// };
+/// use xml_struct::{Error, XmlSerialize};
+///
+/// #[derive(XmlSerialize)]
+/// #[xml_struct(default_ns = "http://foo.example/")]
+/// struct Foo {
+///     some_field: String,
+///
+///     #[xml_struct(flatten)]
+///     something_else: Bar,
+/// }
+///
+/// enum Bar {
+///     Baz,
+///     Qux(String),
+/// }
+///
+/// impl XmlSerialize for Bar {
+///     fn serialize_child_nodes<W>(&self, writer: &mut Writer<W>) -> Result<(), Error>
+///     where
+///         W: std::io::Write,
+///     {
+///         match self {
+///             Self::Baz => writer.write_event(Event::Text(BytesText::new("BAZ")))?,
+///             Self::Qux(qux) => qux.serialize_as_element(writer, "Qux")?,
+///         }
+///
+///         Ok(())
+///     }
+/// }
+///
+/// let mut writer: Writer<Vec<u8>> = Writer::new(Vec::new());
+/// let foo = Foo {
+///     some_field: "foo".into(),
+///     something_else: Bar::Baz,
+/// };
+///
+/// assert!(foo.serialize_as_element(&mut writer, "FlyYouFoo").is_ok());
+///
+/// let out = writer.into_inner();
+/// let out = std::str::from_utf8(&out).unwrap();
+///
+/// assert_eq!(
+///     out,
+///     r#"<FlyYouFoo xmlns="http://foo.example/"><SomeField>foo</SomeField>BAZ</FlyYouFoo>"#,
+/// );
+/// ```
+pub trait XmlSerialize {
+    /// Serializes this value as XML content nodes within an enclosing XML
+    /// element.
+    fn serialize_as_element<W>(&self, writer: &mut Writer<W>, name: &str) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        writer.write_event(Event::Start(BytesStart::new(name)))?;
+
+        self.serialize_child_nodes(writer)?;
+
+        writer.write_event(Event::End(BytesEnd::new(name)))?;
+
+        Ok(())
+    }
+
+    /// Serializes this value as XML content nodes.
+    fn serialize_child_nodes<W>(&self, writer: &mut Writer<W>) -> Result<(), Error>
+    where
+        W: std::io::Write;
+}
+
+/// A data structure which can be serialized as the value of an XML attribute.
+pub trait XmlSerializeAttr {
+    /// Serializes this value as the value of an XML attribute.
+    fn serialize_as_attribute(&self, start_tag: &mut BytesStart, name: &str);
+}
+
+#[derive(Debug, Error)]
+/// An error generated during the XML serialization process.
+pub enum Error {
```

Comment:
Do you intend to support more error types in future ? If so could we decorate with `'#[non_exhaustive]`

Reply:
Ope, thanks for the reminder. Not used to that one.